commisery/versioning.py

```diff
@@ -0,0 +1,463 @@
+    def __ge__(self, rhs: "Version") -> bool: ...
+    # fmt: on
+
+
+class SemVer(Version):
+    """
+    Semantic versioning policy.
+
+    This policy is based on Semantic Versioning 2.0.0: https://semver.org/spec/v2.0.0.html
+
+     * Parsing and serialization is according to the syntax specified by 'semver'.
+     * Version comparison and ordering is implemented exactly as specified by 'semver'.
+     * Version incrementing is implemented exactly as specified by 'semver' for the major, minor and patch fields.
+      - 'semver' doesn't specify a strategy for incrementing of the prerelease field, thus:
+       + given a version B, constructed by incrementing the prerelease field of A
+       + and a version C, constructed by incrementing the prerelease of B
+       + and a version D, constructed by incrementing the patch field of A
+       + we ensure that:
+        * A always sorts before B; and
+        * B always sorts before C; and
+        * C always sorts before D; and
+        * that this relationship is transitive
+    """
+    __slots__ = ('major', 'minor', 'patch', 'prerelease', 'build')
+    default_tag_name = "{version.major}.{version.minor}.{version.patch}{version.prerelease_separator}{version.prerelease}"
+
+    def __init__(self, major: int, minor: int, patch: int, prerelease: Tuple[str, ...], build: Tuple[str, ...]):
+        self.major      = major
+        self.minor      = minor
+        self.patch      = patch
+        self.prerelease = prerelease
+        self.build      = build
+
+    def __setattr__(self, name, value):
+        if name in {'major', 'minor', 'patch'}:
+            return super().__setattr__(name, int(value))
+        elif name in {'prerelease', 'build'}:
+            return super().__setattr__(name, _IdentifierList(value))
+
+    def __iter__(self) -> Iterator[TBD]:
+        return iter(getattr(self, attr) for attr in self.__class__.__slots__)
+
+    def __repr__(self) -> str:
+        return '%s(major=%r, minor=%r, patch=%r, prerelease=%r, build=%r)' % ((self.__class__.__name__,) + tuple(self))
+
+    @property
+    def prerelease_separator(self) -> str:
+        return "-" if self.prerelease else ""
+
+    @property
+    def build_separator(self) -> str:
+        return "+" if self.build else ""
+
+    def __str__(self) -> str:
+        ver = '.'.join(str(x) for x in tuple(self)[:3])
+        ver += f"{self.prerelease_separator}{self.prerelease}"
+        ver += f"{self.build_separator}{self.build}"
+        return ver
+
+    version_re = re.compile(
+        r'^(?:version=)?'
+      + r'(?P<major>0|[1-9][0-9]*)'                                  # noqa: E131
+      + r'\.(?P<minor>0|[1-9][0-9]*)'
+      + r'\.(?P<patch>0|[1-9][0-9]*)'
+      + r'(?:-(?P<prerelease>[-0-9a-zA-Z]+(?:\.[-0-9a-zA-Z]+)*))?'
+      + r'(?:\+(?P<build>[-0-9a-zA-Z]+(?:\.[-0-9a-zA-Z]+)*))?'
+      + r'\s*$'
+    )
+
+    @classmethod
+    def parse(cls, s: str) -> Optional["SemVer"]:
```

Comment:
Might be nice to use `TypeVar` to map a type to the `SemVer` class, to avoid typos when using a string

Reply:
As discussed; this is just an import.